src/engine/Map.cxx

```diff
@@ -757,3 +759,98 @@ void Map::calculateVisibleMap(void)
     }
   }
 }
+
+void Map::setTileID(const std::string &tileID, Point coordinate)
+{
+  TileData *tileData = TileManager::instance().getTileData(tileID);
+  std::vector<Point> targetCoordinates = TileManager::instance().getTargetCoordsOfTile(coordinate, tileID);
+  //TODO: isPlacementOnNodeAllowed should also allow std::vector<Point>, but this is locked until #851 is merged. we can then remove the loop check
+  if (!tileData || targetCoordinates.empty() || !isPlacementOnNodeAllowed(coordinate, tileID))
+  { // if the node would not outside of map boundaries, targetCoordinates would be empty
+    return;
+  }
+
+  for (auto coord : targetCoordinates)
+  { // first check all nodes if it is possible to place the building before doing anything
+    if (!isPlacementOnNodeAllowed(coord, tileID))
+    { //make sure every target coordinate is valid for placement, not just the origin coordinate.
+      return;
+    }
+  }
+
+  Layer layer = TileManager::instance().getTileLayer(tileID);
+  std::string randomGroundDecorationTileID;
+  std::vector<MapNode *> nodesToBeUpdated;
+
+  // if this building has groundDeco, grab a random tileID from the list
+  if (!tileData->groundDecoration.empty())
+  {
+    randomGroundDecorationTileID =
+        *Randomizer::instance().choose(tileData->groundDecoration.begin(), tileData->groundDecoration.end());
+  }
+
+  // for >1x1 buildings, clear all the nodes that are going to be occupied before placing anything.
+  if (targetCoordinates.size() > 1)
+  {
+    demolishNode(targetCoordinates, 0, Layer::BUILDINGS);
+  }
+
+  for (auto coord : targetCoordinates)
+  { // now we can place our building
+
+    MapNode &currentMapNode = mapNodes[nodeIdx(coord.x, coord.y)];
+
+    if (coord != coordinate && targetCoordinates.size() > 1)
+    { // for buildings >1x1 set every node on the layer that will be occupied to invisible exepct of the origin node
+      currentMapNode.setRenderFlag(layer, false);
+    }
+    else
+    { // 1x1 buildings should be set to visible
+      currentMapNode.setRenderFlag(layer, true);
+    }
+
+    if (!targetCoordinates.size() == 1)
```

Comment:
operator precedence of ! is lower than ==
also the note is different of the check, not sure if this was like this before
if it's not a >1x1 building, equals to --> 1x1 building  so the check should be targetCoordinates.size() == 1 without the ! operator, right ?

Reply:
yes, I thought so too, but when I change it to`(!targetCoordinates.size() == 1)` it does this:

<img width="340" alt="image" src="https://user-images.githubusercontent.com/7029836/162631725-7140fc0f-52e2-4ca8-99f9-cd96bc525ae3.png">

no matter if there's a preceding `!` it seems to work correctly.

<img width="770" alt="image" src="https://user-images.githubusercontent.com/7029836/162631801-c1919e52-eca6-410a-941e-1c0c9a5caf76.png">


This function was really hard for me to figure out and to rewrite correctly... maybe I'm not smart enough, but I don't know if I should keep the not at the beginning

Reply:
this is weird